typegen/strats/inline.py

```diff
@@ -1,96 +1,110 @@
-import ast
 import functools
 import logging
 import operator
 import pathlib
-import itertools
+
+import pandas as pd
+
+import libcst as cst
+
+from libcst.metadata import PositionProvider
 
 from constants import TraceData
 from tracing.trace_data_category import TraceDataCategory
 from typegen.strats.gen import TypeHintGenerator
 
-import pandas as pd
-
 logger = logging.getLogger(__name__)
 
 
-class TypeHintTransformer(ast.NodeTransformer):
-    def __init__(self, relevant: pd.DataFrame) -> None:
-        super().__init__()
-        self.df = relevant
+class TargetExtractor(cst.CSTVisitor):
+    targets: list[tuple[str, cst.Name | cst.Attribute]]
 
-    def generic_visit(self, node: ast.AST) -> ast.AST:
-        # Track ClassDefs and FunctionDefs to disambiguate
-        # functions from methods
-        if isinstance(node, ast.ClassDef):
-            for child in filter(lambda c: isinstance(c, ast.FunctionDef), node.body):
-                child.parent = node  # type: ignore
+    def __init__(self):
+        self.targets = list()
 
-        # Track assignments from Functions
-        # NOTE: this handles nested functions too, because the parent reference gets overwritten
-        # NOTE: before we start generating hints for its children
-        if isinstance(node, ast.FunctionDef):
-            for direct in ast.iter_child_nodes(node):
-                for child in ast.walk(direct):  # type: ignore
-                    child.parent = node  # type: ignore
+    def visit_Attribute(self, node: cst.Attribute) -> bool | None:
```

Comment:
Why | None?

Reply:
Because that is the type hint given by the base class `cst.CSTTransformer`